presto-geospatial/src/main/java/com/facebook/presto/plugin/geospatial/GeoFunctions.java

```diff
@@ -1586,7 +1592,218 @@ public static Double stSphericalLength(@SqlType(SPHERICAL_GEOGRAPHY_TYPE_NAME) S
+                (testEdgeStart.getY() < edgeLatitude);
+        return testEdgeStart.getY() < edgeLatitude && inRange;
+    }
+
+    private static double getArcLatitudeAtLongitude(Point start, Point end, double longitude)
+    {
+        // y of longstart must be lower than y of longend
+        Point startUpwardDirection = start;
+        Point endUpwardDirection = end;
+        if (start.getY() > end.getY()) {
+            startUpwardDirection = end;
+            endUpwardDirection = start;
+        }
+        SphericalExcessCalculator calculator = new SphericalExcessCalculator(startUpwardDirection);
+        return calculator.computeExtrapolatedLatitudeThroughPoint(endUpwardDirection, longitude);
+    }
+
+    private static Point getExternalPoint(double longitude)
+    {
+        return new Point(longitude, MAX_LATITUDE);
+    }
+
+    private static int getNumberIntersectionsWithPolygon(Polygon polygon, Point point)
+    {
+        int intersectionSum = 0;
+        Point testEdgeStart = point;
+        Point testEdgeEnd = new Point(point.getX(), MAX_LATITUDE);
+
+        int numPaths = polygon.getPathCount();
+        Boolean excludeEdgeStartPoint = false;
+        for (int i = 0; i < numPaths; i++) {
+            int pathStartIndex = polygon.getPathStart(i);
+            int pathEndIndex = polygon.getPathEnd(i);
+            for (int j = pathStartIndex; j < pathEndIndex - 1; j++) {
+                Point edgeStart = polygon.getPoint(j);
+                Point edgeEnd = polygon.getPoint(j + 1);
+
+                // all vertices are excluded
+                if (edgeStart.equals(point) || edgeEnd.equals(point)) {
+                    return 0;
+                }
+
+                // if this is a vertical edge, it can only add an even (and
+                // therefore uninformative) number of crossings, so don't bother checking
+                if (isVerticalEdge(edgeStart, edgeEnd)) {
+                    excludeEdgeStartPoint = false;
+                    continue;
+                }
+
+                // if the previous edge and the current edge share a vertex,
+                // skip checking it this time to avoid double counting
+                // (if we didn't do this, consider the case of an upside down
+                // triangle with an external point directly below its tip)
+                if (excludeEdgeStartPoint) {
+                    if (haveSameLongitude(edgeStart.getX(), point.getX())) {
+                        continue;
+                    }
+                }
+
+                // the garden variety case
+                if (edgeIntersectsTestEdge(edgeStart, edgeEnd, testEdgeStart, testEdgeEnd)) {
+                    intersectionSum++;
+                }
+
+                excludeEdgeStartPoint = true;
+            }
+            // check last edge which closes the loop back at the starting point
+            Point lastEdgeStart = polygon.getPoint(pathEndIndex - 1);
```

Comment:
I think this logic to handle the last edge can be eliminated by changing the loop to start with the last edge like so. We can also avoid comparing each vertex with input point twice. What do you think?

```
            Point edgeStart = polygon.getPoint(pathEndIndex - 1);
            if (edgeStart.equals(point)) {
                return false;
            }
            
            for (int j = pathStartIndex; j < pathEndIndex; j++) {
                Point edgeEnd = polygon.getPoint(j);

                // all vertices are excluded
                if (edgeEnd.equals(point)) {
                    return false;
                }

                // if this is a vertical edge, it can only add an even (and
                // therefore uninformative) number of crossings, so don't bother checking
                if (isVerticalEdge(edgeStart, edgeEnd)) {
                    excludeEdgeStartPoint = false;
                    edgeStart = edgeEnd;
                    continue;
                }

                // if the previous edge and the current edge share a vertex,
                // skip checking it this time to avoid double counting
                // (if we didn't do this, consider the case of an upside down
                // triangle with an external point directly below its tip)
                if (excludeEdgeStartPoint) {
                    if (haveSameLongitude(edgeStart.getX(), point.getX())) {
                        edgeStart = edgeEnd;
                        continue;
                    }
                }

                // the garden variety case
                if (edgeIntersectsTestEdge(edgeStart, edgeEnd, point, NORTH_POLE)) {
                    intersectionSum++;
                }

                excludeEdgeStartPoint = true;
                edgeStart = edgeEnd;
            }
```